Add button to record decoded audio.

tmiw opened the pull request:
This PR resolves #1091 by adding an additional button on the main window to record decoded audio from the RADE modem:

<img width="189" height="212" alt="image" src="https://github.com/user-attachments/assets/b9770998-d6e1-4df0-9f4e-d792c3238bd6" />

("Record" was also renamed to "Raw" to help differentiate between the two recording types.)

This creates a file named "FreeDV_FromDecoder_..." in the same location as configured for raw recordings (in Tools->Options->Audio->Quick Record).

barjac commented:
This works fine however I think that different paths need to be configurable for each to avoid confusion.

Some users will not fully understand that Raw must be used for playback through FreeDV and that FromDecoder can be used to replay with any audio player or on air using the Voice Keyer.

Using different paths would help to avoid them being mixed up.

It would also be useful to have both buttons show a small dialogue to allow the addition of a short text to the filename e.g. callsign etc. replacing any spaces with '_', for easy identification of the files later.

Filenames could be shorter like:
FDV_Dec_20251226-122648.wav
FDV_Raw_20251226-122648.wav

...and with text added:
FDV_Dec_G2ZZZ_20251226-122648.wav
FDV_Raw_G2ZZZ_20251226-122648.wav

My temporary stand-alone program had this and it was really useful having a callsign in the filename.

barjac commented:
> Hi, Thanks.  I should have been more specific.  Is this feature available in a development release for Windows?

I don't know how dev releases for Win are decided, Mooneer will no doubt comment on that :)

tmiw commented:
> Some users will not fully understand that Raw must be used for playback through FreeDV and that FromDecoder can be used to replay with any audio player or on air using the Voice Keyer.

Should the buttons be renamed to help with that as well? If so, any suggestions?

> With that in mind it would be really useful to be able to select .mp3 as an optional output format so decoded recordings could be sent by email when needed, as .wav are generally too large. Obviously not for replay with the VK.

I think libsndfile can do this, but I might need to investigate further (in case it needs additional dependencies that aren't necessarily available in the Windows or macOS build environment).

tmiw commented:
> Hi, OK.  Please see the attached screen shot.   What am I missing?  I don't see an option to record decoded audio. 73, Ron / W4MMP

Can you log in with a web browser and attach the screenshot there? It's not showing up for me so I think email stripped it out.

tmiw commented:
> Here's the screen shot. [2025-12-27_144159.zip](https://github.com/user-attachments/files/24357843/2025-12-27_144159.zip)

Thanks. Looks like the commit ID is incorrect--try [this](https://github.com/drowe67/freedv-gui/actions/runs/20495426245/artifacts/4964102436)?

barjac commented:
> So the original reason for the Record button in the main window is to provide a quicker method of starting a recording from the radio. In the past, you'd go to Tools->Start Record File - From Radio to provide additional parameters (e.g. filename and the number of seconds to record). That's why I'm fairly hesitant to add additional steps to the record process. I can see if I can shrink the amount of space needed for the buttons, as well as add a Tools menu option for additional recording options.

All understood but the old original way was really clunky.

This seems to be a retrograde step as regards screen use, now in width.

I envisaged something like this mock-up on pressing the Record button with the defaults shown:


<img width="597" height="416" alt="Screenshot_20251229_152504" src="https://github.com/user-attachments/assets/fc421ac0-8f76-4dbe-a38b-a5d9fd686077" />

It is only one more click for instant raw recording but offers so much more, 'almost' instantly.

Anything requiring a trip into the Tools menu is a real turn-off for many users.

tmiw commented:
OK, I did the following as of 8a377ed:

1. Removed Tools->Start Record File... options from the menu bar (Record button is the only way to start recordings).
2. Went back to a single Record button. This button pops up something similar to what the old Tools->Start Record File... options did, just with a different set of controls on the bottom for raw/decoded recording (raw is default). Example (macOS):

<img width="992" height="560" alt="image" src="https://github.com/user-attachments/assets/edceb1fd-7474-46f6-a105-ea6604e13763" />

3. Removed Quick Record config options from Tools->Options since we're now using the method the old Tools menu options used to preserve the last used folder location.

Hopefully this works for you @barjac?

tmiw commented:
> I think one extra check box for optional mp3 creation in the dialogue would finish this off nicely! mp3 could use the same file name and directory with the mp3 extension.

Done as of [0688656](https://github.com/drowe67/freedv-gui/pull/1145/commits/06886567b48b8ecd548d4c23cb5a737e34b60f26).

> Also maybe 'Off Air' rather than 'Raw' has been suggested, and maybe 'Start' rather than 'OK'?

And done.

barjac commented:
This seems to work fine but I have some questions:
1. I assume that 'Off-air' raw .wav is only intended for later replay through the FreeDV decoder, as I can't think of another use for a raw recording? I was thinking that maybe mp3 should be greyed when Off-air is selected, however there is the case where an SSB recording may be wanted for comparison with FreeDV so maybe it should stay?
2. Is mp3 actually suitable for voice keyer use? I discovered that it does transmit OK but I have not yet had any feedback on the received results. If this is acceptable then maybe the VK file chooser should not default to only .wav when launched?
3. Does the recording of mp3 create a .wav then convert it, or is it compressed on the fly?
  If the former then by checking both, could the .wav be retained rather than deleted, or alternatively could they both be created at the same time?
4. I tried to record the decoded output of a file 'played from radio' but the new recording was silent. In my external program I could do that to get both sides of the conversation into an mp3. Is this not possible within FreeDV?

Some explanation will be needed in tooltips for all the options in the dialogue.
After you answer the above I will be happy to prepare some text suggestions for these.

barjac commented:
Re. 1 OK
Re. 2 Could the file chooser please default to *.* at least for now while we do some testing?
I know it can be done manually but it's just another step that gets in the way.
Re. 3 OK understood.
Re. 4 OK re-tested with both wav and mp3 and it is OK.

One minor issue I came across is that if you are in transmit and start voice keyer to play a file over the air, the file plays OK but the plot is not shown in from mic. Starting VK from cold does show in the from mic.

barjac commented:
You misunderstood!
Its good that it defaults to the last used folder used from FreeDV but it currently filters by default to .wav only, so .mp3 are not visible. It would be better if it could default to showing both .wav and .mp3 or 'All files' if that is not possible.

tmiw commented:
> Yes the VK file selector. Its good that it defaults to the last used folder used from FreeDV but it currently filters by default to .wav only, so .mp3 are not visible. It would be better if it could default to showing both .wav and .mp3 or if that is not possible to 'All files'.
> 
> Also would it be possible to allow this file selection to be done while in transmit?
> 
> Also a really neat addition would be to auto-fill the filename suffix field with a callsign IF a row in either the pull-down or reporter was selected prior to hitting the record button. Similar to logging.

All done as of [8264eda](https://github.com/drowe67/freedv-gui/pull/1145/commits/8264eda9fc717e1f919b0b5a62c8a99a31c1406a).
